okhttp/src/test/java/okhttp3/JSSETest.kt

```diff
@@ -0,0 +1,217 @@
+        }
+      }))
+      .build()
+
+    val request = Request.Builder().url("https://facebook.com/robots.txt").build()
+
+    client.newCall(request).execute().use {
+      assertThat(it.protocol).isEqualTo(Protocol.HTTP_2)
+      assertThat(it.handshake!!.tlsVersion).isEqualTo(TlsVersion.TLS_1_3)
+    }
+
+    client.connectionPool.evictAll()
+    assertEquals(0, client.connectionPool.connectionCount())
+
+    client.newCall(request).execute().use {
+      assertThat(it.protocol).isEqualTo(Protocol.HTTP_2)
+      assertThat(it.handshake!!.tlsVersion).isEqualTo(TlsVersion.TLS_1_3)
+    }
+
+    assertEquals(2, sessionIds.size)
+    assertNotEquals(sessionIds[0], sessionIds[1])
+    assertThat(sessionIds[0]).isNotBlank()
+  }
+
+  @Test
+  fun testBuildIfSupported() {
+    val actual = OpenJSSEPlatform.buildIfSupported()
+    assertThat(actual).isNotNull
+  }
+
+  @ParameterizedTest(name = "{displayName}({arguments})")
+  @ValueSource(strings = ["TLSv1.2", "TLSv1.3"])
+  @Flaky
+  fun testSessionReuse(tlsVersion: String) {
+    val sessionIds = mutableListOf<String>()
+
+    enableTls()
+
+    val tlsVersion = TlsVersion.forJavaName(tlsVersion)
+    val spec = ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
+      .tlsVersions(tlsVersion)
+      .build()
+
+    var reuseSession = false
+
+    val sslContext = handshakeCertificates.sslContext()
+    val systemSslSocketFactory = sslContext.socketFactory
+    val sslSocketFactory = object : DelegatingSSLSocketFactory(systemSslSocketFactory) {
+      override fun configureSocket(sslSocket: SSLSocket): SSLSocket {
+        return sslSocket.apply {
+          if (reuseSession) {
+            this.enableSessionCreation = false
+          }
+        }
+      }
+    }
+
+    client = client.newBuilder()
+      .connectionSpecs(listOf(spec))
+      .eventListenerFactory(clientTestRule.wrap(object : EventListener() {
+        override fun connectionAcquired(call: Call, connection: Connection) {
+          val sslSocket = connection.socket() as SSLSocket
+
+          sessionIds.add(sslSocket.session.id.toByteString().hex())
+        }
+      }))
+      .sslSocketFactory(sslSocketFactory, handshakeCertificates.trustManager)
+      .build()
+
+    server.enqueue(MockResponse().setBody("abc1"))
+    server.enqueue(MockResponse().setBody("abc2"))
+
+    val request = Request.Builder().url(server.url("/")).build()
+
+    client.newCall(request).execute().use { response ->
+      assertEquals(200, response.code)
+    }
+
+    client.connectionPool.evictAll()
+    assertEquals(0, client.connectionPool.connectionCount())
+
+    // Force reuse. This appears flaky (30% of the time) even though sessions are reused.
+    // javax.net.ssl.SSLHandshakeException: No new session is allowed and no existing
+    // session can be resumed
+    // TODO: raise JDK bug.
+    // reuseSession = true
+
+    client.newCall(request).execute().use { response ->
+      assertEquals(200, response.code)
+    }
```

Comment:
Extract a helper function that makes two connections in sequence and returns the session IDs? DRY this up.

Reply:
Made a separate test instead